packages/core/src/Site/index.js

```diff
@@ -1467,6 +1482,84 @@ class Site {
     }
   }
 
+  /**
+   * Archive the current version of the site.
+   *
+   * @param {string} versionName the name of the version
+   * @param {string} archivePath the path to the folder to store the archives in
+   */
+  async archive(versionName, archivePath) {
+    await this.readSiteConfig();
+
+    // Save version data
+    const versionPath = `${archivePath}/${versionName}`;
+    this.versionData = await this.writeVersionsFile(versionName, versionPath);
+    // Exclude versioned files from archiving.
+    const archiveFolders = this.versionData.versions;
+    archiveFolders.forEach((folder) => {
+      const filePath = `/${folder.output}/**`;
+      this.siteConfig.ignore.push(filePath);
+    });
+
+    // Used to get accurate intralinks within the archived site:
+    const archivedBaseUrl = this.siteConfig.baseUrl === ''
+      ? `/${versionPath}`
+      : `${this.siteConfig.baseUrl}/${versionPath}`;
+    this.siteConfig.baseUrl = archivedBaseUrl;
```

Comment:
Mostly looks like on the right track, I think the biggest missing thing may be:

> copy over the generated version of the website into a sub folder under the _sites folder for subsequent deployment

I realised with this note, perhaps we could also store the `baseUrl` used to archive the version (if we ever want to support changing baseurls) inside `versions.json`, as differing baseUrls shouldn't be copied over to `_site`.

Reply:
I'm not sure what you mean by "differing baseUrls sholudn't be copied over to `_site`" - from my understanding, the HTML files do away with references to baseURL by replacing references to baseURL with the baseURL. But yes, I can definitely store the archived site's baseUrl!

Reply:
Hmm, I see...

It makes sense that differing baseUrls shouldn't be copied over (as they wouldn't be able to be deployed), but  I wonder if there are any cases where the baseUrls might differ, but you might still want to include it as part of the same deployment and it would work (something like a root baseUrl being 'root' and the subsite baseUrl being "root/subsite" - it still could be part of the same deployment (?)).

I think copying it over is not a significant cost or a bug, and if the user wants to it is easy to manually exclude by indicating that folder as a folder to ignore. But it shouldn't be too hard to exclude the files as well 🤔

Reply:
> but I wonder if there are any cases where the baseUrls might differ, but you might still want to include it as part of the same deployment and it would work (something like a root baseUrl being 'root' and the subsite baseUrl being "root/subsite" - it still could be part of the same deployment (?)).

Subsites are meant to be entire sites able to be deployed independently.

Our support for subsites is more about content reuse https://markbind.org/userGuide/reusingContents.html#reusing-contents-across-sites, in particular revolving around resolving `{{baseUrl}}` properly

E.g.

Main site, baseUrl is '/cat'
Subsite , baseUrl is '/dog'

subsite index.md
```
[some link]({{baseUrl}}/foobar)
```

If you deploy the subsite independently, `baseUrl` resolves to `/dog`.

If you deploy the main site, `baseUrl` resolves to the main site base url + **sub folder** (i.e. `/cat/subfolder/of/subsite`).

This is important wherever there are links involved (link validaiton, relative link resolving, `<include>`s, ...).

> by indicating that folder as a folder to ignore.

If I'm getting this part right (ignore as in the `ignore` property), the versioning system's copying mechanism needs to be independent of asset copying (the `ignore` property).

e.g. if you have

```
ignore: ["xx.png"]
```

but at the point of archiving, `xx.png` wasn't present (meaning the author did indeed want to copy it over). The copying for versioning should still copy `xx.png` over.

Reply:
> at the point of archiving, `xx.png` wasn't present (meaning the author did indeed want to copy it over). The copying for versioning should still copy `xx.png` over.

Yes, this is currently accounted for. At the point of archiving, site.json is checked to save the archived files, and is not updated unless you archive a newer version under the same name. So if `ignore: ["xx.png"]` is added afterwards, the archived site would still have `xx.png` but the current site would not.

Reply:
> the versioning system's copying mechanism needs to be independent of asset copying (the ignore property).

> The copying for versioning should still copy xx.png over.

Referring to `markbind build` copying (not `archive`), which hasn't been implemented yet.

Reply:
(sorry I don't understand, could you explain that again?)

Reply:
> Mostly looks like on the right track, I think the biggest missing thing may be:
> 
> > copy over the generated version of the website into a sub folder under the _sites folder for subsequent deployment

same part here ^ @kaixin-hc

Reply:
From what I can see currently this hasn't been implemented, so I assume you're relying on asset copying / `ignore` property to this. Apologise as I think it might be clearer if I just assumed it hadn't been implemented 😅

Reply:
The point on `ignore` however is this: If you had archived a site, say into `versions/v1`. Then, you ran `markbind build/serve`. MarkBind's asset copying would incorrectly / incompletely copy some parts of it over (since it is in the project source folder). This copying mechanism therefore needs to be implemented separately of this. The asset copying should also exclude said archive folders.

Reply:
Oh :0 So the desired default behaviour, in your opinion, is to serve only the current site, and not any archived sites, because of the cost in copying over large sites? That makes sense. I had assumed the opposite, which is that if the user had archived the site previously they would want to check it when building and serving.

Currently, my understanding/implementation is:
 
* When the site is archived, the site is built with a changed baseUrl and placed into a certain folder, the location of which is tracked in `versions.json` - previously archived versions will not be re-archived.
    * (Currently working on subsite versions not being re-archived, should be done soon) 
* Every time the site is built/served, the contents of `_site` are wiped clean, and replaced by the newly built files + copied over assets (which now include the versioned sites).
    * To avoid deploying versioned sites, this is currently relying on the user manually specifying to ignore that folder.
    * Asset copying "correctly" moves over all the versioned sites, allowing the user to navigate to past versions in their live preview 
* Every time the site is deployed, the contents of `_site` and only those contents are transferred to `gh-pages` branch --> this will include all versioned sites which were not specifically ignored by the user.

And what I understand from your suggestion is that:
* when archiving, also move the archived site files into a subfolder in `_site`
* change the way the site is built and served such that only the current site is built and served, ignoring the new subfolders in `_site` which hold versions.
    * Hence we can't wipe `_site` clean, and will need to store the names of subfolders to "ignore"(either not overriding with versioning or not serving with markbind serve)
    * And I assume also adding options to allow the build and serve commands to also show archived sites, if specified. 
(One clarifying question: does this mean you suggest to store the archived site two places, both in `_site` and in the folder `<archivePath>/<versionName>` ?)

I can give it a shot, but do you think I can leave this as an improvement for another PR? I think it might be a little tricky because it affects big commands and features and might take more than a few days to implement carefully (might also make reviewing easier)

Reply:
Nope, nothing so complex, ignore the additional point on `markbind serve` first, I'll come back to that later.
Your current understanding is mostly already correct.

> and replaced by the newly built files + copied over assets (which now include the versioned sites).

> Asset copying "correctly" moves over all the versioned sites, allowing the user to navigate to past versions in their live preview

This is the main issue, it is _incorrect to rely on asset copying_ for this. https://github.com/MarkBind/markbind/pull/1870#discussion_r851633960

You'll have to implement an independent copying mechanism.

If you need convincing, try archiving our documentation with `markbind archive v1`. Then, add `appveyorAddNewProject.png` to `ignore`, and run `markbind build`. The image will be missing from the archived site in `versions/v1/images`.

Reply:
*Please make sure you get the above comment before reading this comment, otherwise it'll only be more confusing

> ~in which case (for `markbind serve`), I think to avoid copying the entire archived site over every file change/add/delete (expensive to copy large site over), we could also aim to do this only once 🤔 ~(realised its not possible to preview archived sites in `markbind serve` with https://github.com/MarkBind/markbind/issues/267 not done yet)~ (might be possible)

Back to this, my concern here was just with this:

1. author runs `markbind serve`
2. your yet-to-implement copying mechanism copies the versioned sites over
3. author edits files... continues with development
  3.1 Triggers your copying mechanism _everytime_ (file change/addition/deletion). This is rather disk intensive / expensive and might slow the development experience a fair bit.

> change the way the site is built and served such that only the current site is built and served, ignoring the new subfolders in _site which hold versions.

Not in the way you are thinking, but I realised with your comment you'll also need to add the _versioned site folders_ to the `pagesExclude` property here.
https://markbind.org/userGuide/siteJsonFile.html#pagesexclude

This is as:
- we support building html files as pages as well
- `.md` files may be copied to output folders as an asset in some projects

Reply:
Thanks for explaining that again, and for the simple example - that's really useful to test my thoughts.

I hadn't realized that we "supported building html files as pages". I had the understanding that "building" a page was converting it from `.md` to a generated `.html` - will take a closer look at this.